File: r/sum-of-multiples/sum-of-multiples.R
library(magrittr)


sum_of_multiples <- function(factors, limit) {
  ### Take 1, iterative
  #  multiples <- c()
  #  # discard any factors that are not smaller than the limit
  #  for (f in Filter(function(f) f < limit, factors)) {
  #    multiples <- append(multiples, seq(f, limit - 1, by = f))
  #  }
  #  sum(unique(multiples))
  ###

  # Take 2, more functional

  # get the multiples of a factor that are less than the limit
  multiples <- function(f) seq(f, limit - 1, by = f)

  # discard any factors that are not smaller than the limit
  Filter(function(f) f < limit, factors) %>%
    sapply(multiples) %>%
    unlist            %>%
    unique            %>%
    sum
}

File: julia/sum-of-multiples/sum-of-multiples.jl
function sum_of_multiples(limit, factors)
	fs          = filter(!iszero, factors)
	is_multiple = i -> any(f -> i % f == 0, fs)
	multiples   = a -> filter(is_multiple, a)

	1:limit-1 |> multiples |> sum
end
